src/jlgen.jl

```diff
@@ -62,15 +62,20 @@ methodinstance
 if VERSION >= v"1.11.0-DEV.1552"
 
 # XXX: version of Base.method_instance that uses a function type
-function methodinstance(ft, tt, world=tls_world_age())
+@inline function methodinstance(@nospecialize(ft::Type), @nospecialize(tt::Type), world::Integer=tls_world_age())
     sig = signature_type_by_tt(ft, tt)
 
     mi = ccall(:jl_method_lookup_by_tt, Any,
                (Any, Csize_t, Any),
                sig, world, #=method_table=# nothing)
     mi === nothing && throw(MethodError(ft, tt, world))
-
-    return mi::MethodInstance
+    mi = mi::MethodInstance
+    # `jl_method_lookup_by_tt` and `jl_method_lookup` can return a unspecialized mi
+    if mi.specTypes !== sig
+        mi = CC.specialize_method(mi.def, sig, Core.svec(), preexisting=true)
+        mi === nothing && throw(MethodError(ft, tt, world))
```

Comment:
AFAIK, the sparams is not optional here, unless you enjoy corrupt caches leading to memory corruption
```suggestion
        sparams = ...
        mi = CC.specialize_method(mi.def, sig, sparams)::MethodInstance
```
(someday I want it to be optional, and computed on-demand if empty, but that day is probably not today)

Reply:
I was looking at uses of `specialize_method` in Base and seen a couple of places where it uses `sparams=Core.svec()`

I don't think I actually know  what `sparams` ought to be.

Reply:
Could I use `mi.sparams_vals` e.g. reuse the sparams from the unspecialized mi?

Reply:
yeah, that seems probably true